bokehjs/src/lib/client/connection.ts

```diff
@@ -214,47 +186,34 @@ export class ClientConnection {
       this._close_bad_protocol(e.toString())
     }
 
-    if (this._receiver.message == null)
-      return
-
     const msg = this._receiver.message
+    if (msg != null) {
+      const problem = msg.problem()
+      if (problem != null)
+        this._close_bad_protocol(problem)
 
-    const problem = msg.problem()
-    if (problem != null)
-      this._close_bad_protocol(problem)
-
-    this._current_handler!(msg)
+      this._current_handler!(msg)
+    }
   }
 
-  protected _on_close(event: CloseEvent): void {
+  protected _on_close(event: CloseEvent, reject: Rejecter): void {
     logger.info(`Lost websocket ${this._number} connection, ${event.code} (${event.reason})`)
     this.socket = null
 
-    if (this._pending_ack != null) {
-      this._pending_ack[1](new Error(`Lost websocket connection, ${event.code} (${event.reason})`))
-      this._pending_ack = null
-    }
+    Object.values(this._pending_replies).forEach(([, reject]) => reject("Disconnected"))
```

Comment:
After switching to `Map<K, V>`, this can be changed to `for (const [, reject] of this._pending_replies) ...`.

Reply:
I would still have to write `this._pending_replies.values()`. But the map, `forEach` won't require calling `Object.values`.

Reply:
There are two options, either use `.values()`:
```ts
for (const v of this._pending_replies.values()) {
  ...
}
```
or 
```ts
for (const [, v] of this._pending_replies) {
  ...
}
```
which is equivalent of using `.items()`. This is due to `Symbol.iterator` support:
```ts
interface Map<K, V> {
    /** Returns an iterable of entries in the map. */
    [Symbol.iterator](): IterableIterator<[K, V]>;
}
```

Reply:
With the latest commit, I switched to `this._pending_replies.forEach`. It seems like the most clean and clear version. And also the shortest one.

Reply:
That's fine, though `forEach` has a peculiar order of arguments when it comes to the `Map` type. However, all those "functional" methods, `forEach`, `map`, `filter`, etc., are also usually the slowest option, so if performance is at stake, then one has to be wary of this choice. In this case it shouldn't matter though.